lib/pengine/pe_actions.c

```diff
@@ -507,15 +510,21 @@ validate_on_fail(const pe_resource_t *rsc, const char *action_name,
             }
 
             // Use value from first applicable promote action found
-            return on_fail;
+            key = strdup(XML_OP_ATTR_ON_FAIL);
+            new_value = strdup(promote_on_fail);
+            CRM_ASSERT((key != NULL) && (new_value != NULL));
```

Comment:
We can use `pcmk__str_update()` in place of both `strdup()` calls, and that includes a `CRM_ASSERT()`.